File: JWTDecoder.kt
// JWTDecoder.kt
import java.io.File
import java.nio.file.Files
import java.nio.file.Paths
import java.util.*
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec
import com.google.gson.GsonBuilder
import com.google.gson.JsonObject
import com.google.gson.JsonParser

class JWTDecoder(private val token: String, private val key: String?) {
    private var header: JsonObject? = null
    private var payload: JsonObject? = null
    private var signature: String? = null
    private var validSignature: Boolean? = null
    private var expired: Boolean? = null

    fun decode() {
        val parts = token.split(".")
        if (parts.size != 3) throw IllegalArgumentException("Invalid JWT format")

        val headerJson = String(Base64.getUrlDecoder().decode(parts[0]))
        header = JsonParser.parseString(headerJson).asJsonObject

        val payloadJson = String(Base64.getUrlDecoder().decode(parts[1]))
        payload = JsonParser.parseString(payloadJson).asJsonObject

        signature = parts[2]

        // Check exp
        if (payload!!.has("exp")) {
            val exp = payload!!.get("exp").asLong
            expired = System.currentTimeMillis() / 1000 > exp
        }

        // Verify signature
        if (key != null && key.isNotEmpty()) {
            validSignature = verifySignature(parts)
        }
    }

    private fun verifySignature(parts: List<String>): Boolean {
        val data = "${parts[0]}.${parts[1]}"
        val mac = Mac.getInstance("HmacSHA256")
        val secretKey = SecretKeySpec(key!!.toByteArray(), "HmacSHA256")
        mac.init(secretKey)
        val sigBytes = mac.doFinal(data.toByteArray())
        val sigB64 = Base64.getUrlEncoder().withoutPadding().encodeToString(sigBytes)
        return sigB64 == parts[2]
    }

    fun toMap(): Map<String, Any?> {
        val map = mutableMapOf<String, Any?>()
        map["header"] = header
        map["payload"] = payload
        map["signature"] = signature
        map["token"] = token
        if (validSignature != null) map["valid_signature"] = validSignature
        if (expired != null) map["expired"] = expired
        return map
    }

    fun print(color: Boolean) {
        val gson = GsonBuilder().setPrettyPrinting().create()
        println("Header:")
        println(gson.toJson(header))
        println("Payload:")
        println(gson.toJson(payload))
        if (expired != null) {
            println("  Token ${if (expired!!) "expired" else "valid"}")
        }
        if (validSignature != null) {
            println("  Signature ${if (validSignature!!) "verified" else "invalid"}")
        }
        println("Signature: $signature")
    }
}

fun main(args: Array<String>) {
    var token: String? = null
    var key: String? = null
    var file: String? = null
    var jsonOut = false
    var color = false

    var i = 0
    while (i < args.size) {
        when (args[i]) {
            "--key" -> key = args[++i]
            "--file" -> file = args[++i]
            "--json" -> jsonOut = true
            "--color" -> color = true
            else -> if (token == null) token = args[i]
        }
        i++
    }

    if (file != null) {
        token = File(file).readText().trim()
    }
    if (token == null) {
        // stdin
        if (System.console() == null) {
            token = Scanner(System.`in`).nextLine().trim()
        } else {
            System.err.println("Error: no token provided")
            System.exit(1)
        }
    }

    val decoder = JWTDecoder(token!!, key)
    try {
        decoder.decode()
    } catch (e: Exception) {
        System.err.println("Error: ${e.message}")
        System.exit(1)
    }

    if (jsonOut) {
        val gson = GsonBuilder().setPrettyPrinting().create()
        println(gson.toJson(decoder.toMap()))
    } else {
        decoder.print(color || System.console() != null)
    }
}

File: JWTDecoder.cs
// JWTDecoder.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;
using System.Security.Cryptography;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

class JWTDecoder
{
    private string token;
    private string key;
    private JObject header;
    private JObject payload;
    private string signature;
    private bool? validSignature;
    private bool? expired;

    public JWTDecoder(string token, string key)
    {
        this.token = token;
        this.key = key;
    }

    public void Decode()
    {
        var parts = token.Split('.');
        if (parts.Length != 3) throw new ArgumentException("Invalid JWT format");

        // Decode header
        var headerJson = Encoding.UTF8.GetString(Convert.FromBase64String(PadBase64(parts[0])));
        header = JObject.Parse(headerJson);

        // Decode payload
        var payloadJson = Encoding.UTF8.GetString(Convert.FromBase64String(PadBase64(parts[1])));
        payload = JObject.Parse(payloadJson);

        signature = parts[2];

        // Check exp
        if (payload.ContainsKey("exp"))
        {
            long exp = payload.Value<long>("exp");
            expired = DateTimeOffset.UtcNow.ToUnixTimeSeconds() > exp;
        }

        // Verify signature
        if (!string.IsNullOrEmpty(key))
        {
            validSignature = VerifySignature(parts);
        }
    }

    private string PadBase64(string base64)
    {
        int padLength = (4 - base64.Length % 4) % 4;
        return base64 + new string('=', padLength);
    }

    private bool VerifySignature(string[] parts)
    {
        var data = parts[0] + "." + parts[1];
        using (var hmac = new HMACSHA256(Encoding.UTF8.GetBytes(key)))
        {
            var sigBytes = hmac.ComputeHash(Encoding.UTF8.GetBytes(data));
            var sigB64 = Convert.ToBase64String(sigBytes)
                .Replace('+', '-').Replace('/', '_').TrimEnd('=');
            return sigB64 == parts[2];
        }
    }

    public Dictionary<string, object> ToDict()
    {
        var dict = new Dictionary<string, object>
        {
            ["header"] = header,
            ["payload"] = payload,
            ["signature"] = signature,
            ["token"] = token
        };
        if (validSignature.HasValue) dict["valid_signature"] = validSignature.Value;
        if (expired.HasValue) dict["expired"] = expired.Value;
        return dict;
    }

    public void Print(bool color)
    {
        Console.WriteLine("Header:");
        Console.WriteLine(header.ToString(Formatting.Indented));
        Console.WriteLine("Payload:");
        Console.WriteLine(payload.ToString(Formatting.Indented));
        if (expired.HasValue)
        {
            Console.WriteLine($"  Token {(expired.Value ? "expired" : "valid")}");
        }
        if (validSignature.HasValue)
        {
            Console.WriteLine($"  Signature {(validSignature.Value ? "verified" : "invalid")}");
        }
        Console.WriteLine($"Signature: {signature}");
    }

    static void Main(string[] args)
    {
        string token = null, key = null, file = null;
        bool jsonOut = false, color = false;

        for (int i = 0; i < args.Length; i++)
        {
            switch (args[i])
            {
                case "--key": key = args[++i]; break;
                case "--file": file = args[++i]; break;
                case "--json": jsonOut = true; break;
                case "--color": color = true; break;
                default: if (token == null) token = args[i]; break;
            }
        }

        if (file != null)
        {
            token = File.ReadAllText(file).Trim();
        }
        if (token == null)
        {
            // stdin
            if (Console.IsInputRedirected)
            {
                token = Console.In.ReadToEnd().Trim();
            }
            else
            {
                Console.Error.WriteLine("Error: no token provided");
                Environment.Exit(1);
            }
        }

        var decoder = new JWTDecoder(token, key);
        try { decoder.Decode(); }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"Error: {ex.Message}");
            Environment.Exit(1);
        }

        if (jsonOut)
        {
            Console.WriteLine(JsonConvert.SerializeObject(decoder.ToDict(), Formatting.Indented));
        }
        else
        {
            decoder.Print(color || Console.IsOutputRedirected == false);
        }
    }
}
